pkg/executor/statefulset.go

```diff
@@ -171,29 +171,103 @@ func UpdateStatefulSetNonReplicas(
 		needPatch = true
 	}
 
-	// EZML-862
+	// https://github.com/bluek8s/kubedirector/issues/229
 	// We should also compare the current statefulset container image against
-	// the image defined for the corresponding role in the KDApp spec. If images
-	// are not the same, KDApp spec image has a higher priority, so we shoud
+	// the image defined for the corresponding role in the KDCluster spec. If images
+	// are not the same, KDCluster spec image has a higher priority, so we should
 	// upgrade the statefulset one.
 	// Make sure, that according this logic, there is no sence to edit a statefulset
-	// directly, as it will be reconciled back to the KDApp spec state.
+	// directly, as it will be reconciled back to the KDCluster spec state.
 
-	// First, check if KDCluster is in configured state
-	if clusterIsReady {
-		appRoleImage, err := catalog.ImageForRole(cr, role.Name)
+	containers := shared.StatefulSetContainers(statefulSet)
+	currentRoleImage := containers[0].Image
+
+	appRoleImage, err := catalog.ImageForRole(cr, role.Name)
+	if err != nil {
+		return err
+	}
+
+	// There may appear the situation when during image upgrade we set to the
+	// cluster spec an incorrect image tag or image cannot be pulled from a repo by some
+	// other reason. It may cause that a member of the role falls
+	// into an endless restart loop getting ErrImagePull or ImagePullBackOff errors.
+	// Than, the cluster falls into non-ready state and this role status will be always
+	// in upgrading state and reject any try to re-upgrade an image.
+	// For avoiding this situation we should have an ability to re-edit cluster spec
+	// and give a chance to the role status to restart again.
+	// A rollback process will start when a cluster app spec field is equal to the
+	// correcponding RollbackInfo fields, stored for this cluster
+	rollback := false
+
+	// Check, if cluster is not ready and current role status has upgrading members
+	roleStatusIsUpgrading := shared.RoleStatusIsUpgrading(cr, role.Name)
```

Comment:
I'm not sure how deep I should dig into reviewing this at the moment ... cf. my review of the validator, I think there will need to be changes in the validator that will also ripple into the reconciliation here.

Broadly speaking I'd say that if the validator does its job then this handler can be simpler. We should know by the time we get here:
* the spec is valid
* it's OK to do the upgrade or rollback if the spec/status is indicating such an operation
* the kdapp identified in spec.app gives us the target image regardless of whether we're upgrading or rolling back

Reply:
Well, I refactored this part